### ISSUE TITLE
Warn if str.parse if no explicit type specified

### ISSUE THREAD

**Issue by shuoli84** [rust-lang/rust-clippy#9107]
### What it does

Warn for `&str.parse()` if the inferred type is i32 without explicit specify.

### Lint Name

STR_PARSE_INFER_TO_I32

### Category

suspicious

### Advantage

Hint developer that the parse actually only process i32, if the value is larger, the code fails.

### Drawbacks

None

### Example

```rust
x.parse().unwrap_or(0) 
```

Could be written as:

```rust
x.parse::<i32>().unwrap_or(0)
```

```rust
if "111111111111111111".parse().unwrap_or(0) > 0 {
    println!("good");
} else {
    println!("bad");
}

if "111111111111111111".parse::<_>().unwrap_or(0) > 0 {
    println!("good");
} else {
    println!("bad");
}

if "111111111111111111".parse::<i64>().unwrap_or(0) > 0 {
    println!("good");
} else {
    println!("bad");
}
```

outputs

```
bad
bad
good
```

**Comment by Centri3** [comment]
This could probably lint f64 as well

**Comment by shuoli84** [comment]
sorry guys, I've moved to other areas and lost my working memory for this hint. :'(

@rustbot release-assignment
